RobotCrawler/src/main/java/parser/empik/EmpikParser.java

```diff
@@ -19,10 +23,26 @@ public EmpikParser(
         this.bookPageParser = bookPageParser;
     }
 
-    public List<BookData> parse() {
+    public Map<String, List<BookData>> parse() {
         List<String> categoryPageUrls = promotionsPageParser.extractCategoryPageUrls();
         List<String> bookPageUrls = categoryPageParser.extractBookPageUrls(categoryPageUrls);
-        return bookPageParser.extractBooks(bookPageUrls);
+        List<BookData> booksData = bookPageParser.extractBooks(bookPageUrls);
+
+        return booksData.parallelStream().collect(Collector.of(HashMap::new, (map, e) -> {
+            if (!map.containsKey(e.tag)) {
+                map.put(e.tag, new ArrayList<>());
+            }
+
+            map.get(e.tag).add(e);
+        }, (m1, m2) -> {
+            m2.entrySet().stream().forEach(entry -> {
+                if (!m1.containsKey(entry.getKey())) {
+                    m1.put(entry.getKey(), new ArrayList<>());
+                }
+                m1.get(entry.getKey()).addAll(entry.getValue());
+            });
+            return m1;
+        }));
```

Comment:
OK, but it still is very hard to debug in the future. I'm sure you will find a cleaner way for that. You can add `// TODO` to reftor this. It will be more easy(a little slower) to debug, if you divide into 2 methods.